File: examples/typescript/gmbh.ts
import { createWalletClient, http, parseAbi, publicActions } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { base } from 'viem/chains';
import * as dotenv from 'dotenv';
import { randomUUID } from 'crypto';

dotenv.config({ path: '../.env' });

// Configuration
const PRIVATE_KEY = (process.env.PRIVATE_KEY || '0x...') as `0x${string}`;
const RPC_URL = process.env.RPC_URL || 'https://mainnet.base.org';
const API_BASE = process.env.API_BASE || 'https://agentbureau-api.datafortress.cloud/v1';
const USDC_ADDRESS = (process.env.USDC_ADDRESS || '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913') as `0x${string}`;
const SERVICE_ENDPOINT = `${API_BASE}/companies/formations`;

const account = privateKeyToAccount(PRIVATE_KEY);
const client = createWalletClient({
  account,
  chain: base,
  transport: http(RPC_URL),
}).extend(publicActions);

async function runFlow() {
  const payload = {
    company_name: "Autonomous Ventures GmbH",
    company_type: "GMBH",
    stammkapital: 12500.0,
    founder_name: "Alice Agent",
    founder_address: "Digital Ether 0x123",
    description: "Building the future of agentic economy."
  };

  const idempotencyKey = randomUUID();
  console.log(`Calling ${SERVICE_ENDPOINT} with Idempotency-Key: ${idempotencyKey}...`);

  // 1. Initial Call (Expect 402)
  const response = await fetch(SERVICE_ENDPOINT, {
    method: 'POST',
    headers: { 
      'Content-Type': 'application/json',
      'Idempotency-Key': idempotencyKey
    },
    body: JSON.stringify(payload),
  });

  if (response.status === 402) {
    const data = await response.json();
    const intentId = data.intent_id as `0x${string}`;
    const paymentLink = data.payment_link;
    
    console.log(`402 Received. Intent ID: ${intentId}`);

    // Parse payment link for escrow deposit arguments
    // ethereum:<address>@<chain_id>/deposit?bytes32=<intent_id>&uint256=<amount>&uint256=<duration>&address=<payee>&uint16=<feeBps>&uint16[3]=<milestones>&bytes=<sig>
    const url = new URL(paymentLink.replace('ethereum:', 'http:')); // trick URL parser
    const escrowAddress = url.pathname.split('@')[0].replace('/', '') as `0x${string}`;
    const params = url.searchParams;
    
    const amountRaw = BigInt(params.get('uint256') || '0');
    const duration = BigInt(url.search.match(/uint256=([^&]+)/g)?.[1].split('=')[1] || '0'); 
    // Manual parsing because of duplicate keys in URLSearchParams
    const allUint256 = url.search.match(/uint256=([^&]+)/g)?.map(v => v.split('=')[1]) || [];
    const amount = BigInt(allUint256[0] || '0');
    const dur = BigInt(allUint256[1] || '0');
    const payee = params.get('address') as `0x${string}`;
    const feeBps = parseInt(params.get('uint16') || '0');
    const milestones = params.get('uint16[3]')?.split(',').map(Number) as [number, number, number];
    const operatorSignature = params.get('bytes') as `0x${string}`;

    console.log(`Depositing ${Number(amount) / 1e6} USDC into Escrow at ${escrowAddress}...`);

    // 2. Approve USDC for Escrow
    const usdcAbi = parseAbi([
      'function approve(address spender, uint256 amount) returns (bool)'
    ]);
    
    console.log('Approving USDC...');
    const approveHash = await client.writeContract({
      address: USDC_ADDRESS,
      abi: usdcAbi,
      functionName: 'approve',
      args: [escrowAddress, amount],
    });
    await client.waitForTransactionReceipt({ hash: approveHash });

    // 3. Send Escrow Deposit
    const escrowAbi = parseAbi([
      'function deposit(bytes32 intentId, uint256 amount, uint256 duration, address payee, uint16 feeBps, uint16[3] memory milestoneWeights, bytes calldata operatorSignature) external'
    ]);
    
    const hash = await client.writeContract({
      address: escrowAddress,
      abi: escrowAbi,
      functionName: 'deposit',
      args: [intentId, amount, dur, payee, feeBps, milestones, operatorSignature],
    });

    console.log(`Deposit transaction sent: ${hash}. Waiting for confirmation...`);
    await client.waitForTransactionReceipt({ hash });

    // 4. Sign the Intent ID for Agent Authorization
    console.log('Signing intent for authorization...');
    const authorization = await client.signMessage({
      message: `AgentBureau Intent: ${intentId}`,
    });

    // 5. Retry with payment signature and authorization
    console.log('Retrying with payment signature and authorization...');
    const finalResponse = await fetch(SERVICE_ENDPOINT, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Idempotency-Key': idempotencyKey,
        'PAYMENT-SIGNATURE': hash,
        'PAYMENT-AUTHORIZATION': authorization,
      },
      body: JSON.stringify(payload),
    });

    if (finalResponse.ok) {
        console.log('Success!');
        console.log(await finalResponse.json());
    } else {
        console.log(`Final call failed with status ${finalResponse.status}`);
        console.log(await finalResponse.text());
    }
  } else {
    console.log(`Status: ${response.status}`);
    console.log(await response.text());
  }
}

runFlow().catch(console.error);

File: examples/python/gmbh.py
import os
import httpx
import uuid
from web3 import Web3
from eth_account.messages import encode_defunct
from urllib.parse import urlparse, parse_qs
from dotenv import load_dotenv

load_dotenv()

# Configuration
RPC_URL = os.getenv("RPC_URL", "https://mainnet.base.org")
USDC_ADDRESS = os.getenv("USDC_ADDRESS", "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913")
PRIVATE_KEY = os.getenv("PRIVATE_KEY", "your_private_key_here")
API_BASE = os.getenv("API_BASE", "https://agentbureau-api.datafortress.cloud/v1")
SERVICE_ENDPOINT = f"{API_BASE}/companies/formations"

w3 = Web3(Web3.HTTPProvider(RPC_URL))
account = w3.eth.account.from_key(PRIVATE_KEY)

# Payload for GmbH Formation
payload = {
    "company_name": "Autonomous Ventures GmbH",
    "company_type": "GMBH",
    "stammkapital": 12500.0,
    "founder_name": "Alice Agent",
    "founder_address": "Digital Ether 0x123",
    "description": "Building the future of agentic economy."
}

def run_flow():
    idempotency_key = str(uuid.uuid4())
    print(f"Calling {SERVICE_ENDPOINT} with Idempotency-Key: {idempotency_key}...")
    
    # 1. Initial Call (Expect 402)
    headers = {"Idempotency-Key": idempotency_key}
    try:
        response = httpx.post(SERVICE_ENDPOINT, json=payload, headers=headers)
    except Exception as e:
        print(f"Error connecting to API: {e}")
        return

    if response.status_code == 402:
        # Extract details
        data = response.json()
        intent_id = data.get("intent_id")
        payment_link = data.get("payment_link")
        
        print(f"402 Received. Intent ID: {intent_id}")
        
        # Parse payment link for escrow deposit arguments
        # ethereum:<address>@<chain_id>/deposit?bytes32=<intent_id>&uint256=<amount>&uint256=<duration>&address=<payee>&uint16=<feeBps>&uint16[3]=<milestones>&bytes=<sig>
        parsed_link = urlparse(payment_link)
        escrow_address = parsed_link.path.split('@')[0]
        params = parse_qs(parsed_link.query)
        
        amount_raw = int(params['uint256'][0])
        duration = int(params['uint256'][1])
        payee = params['address'][0]
        fee_bps = int(params['uint16'][0])
        milestones = [int(m) for m in params['uint16[3]'][0].split(',')]
        operator_signature = params['bytes'][0]

        print(f"Depositing {amount_raw / 1e6} USDC into Escrow at {escrow_address}...")

        # 2. Approve USDC for Escrow
        usdc_abi = [{"constant":False,"inputs":[{"name":"_to","type":"address"},{"name":"_value","type":"uint256"}],"name":"transfer","outputs":[{"name":"","type":"bool"}],"type":"function"},
                    {"constant":False,"inputs":[{"name":"_spender","type":"address"},{"name":"_value","type":"uint256"}],"name":"approve","outputs":[{"name":"","type":"bool"}],"type":"function"}]
        usdc_contract = w3.eth.contract(address=USDC_ADDRESS, abi=usdc_abi)
        
        # Approve
        approve_tx = usdc_contract.functions.approve(escrow_address, amount_raw).build_transaction({
            'chainId': 8453,
            'gas': 100000,
            'gasPrice': w3.eth.gas_price,
            'nonce': w3.eth.get_transaction_count(account.address),
        })
        signed_approve = w3.eth.account.sign_transaction(approve_tx, PRIVATE_KEY)
        w3.eth.send_raw_transaction(signed_approve.raw_transaction)
        print("USDC Approval sent.")

        # 3. Send Escrow Deposit
        escrow_abi = [{"inputs":[{"internalType":"bytes32","name":"intentId","type":"bytes32"},{"internalType":"uint256","name":"amount","type":"uint256"},{"internalType":"uint256","name":"duration","type":"uint256"},{"internalType":"address","name":"payee","type":"address"},{"internalType":"uint16","name":"feeBps","type":"uint16"},{"internalType":"uint16[3]","name":"milestoneWeights","type":"uint16[3]"},{"internalType":"bytes","name":"operatorSignature","type":"bytes"}],"name":"deposit","outputs":[],"stateMutability":"nonpayable","type":"function"}]
        escrow_contract = w3.eth.contract(address=escrow_address, abi=escrow_abi)
        
        deposit_tx = escrow_contract.functions.deposit(
            intent_id,
            amount_raw,
            duration,
            payee,
            fee_bps,
            milestones,
            operator_signature
        ).build_transaction({
            'chainId': 8453,
            'gas': 300000,
            'gasPrice': w3.eth.gas_price,
            'nonce': w3.eth.get_transaction_count(account.address),
        })
        
        signed_deposit = w3.eth.account.sign_transaction(deposit_tx, PRIVATE_KEY)
        tx_hash = w3.eth.send_raw_transaction(signed_deposit.raw_transaction).hex()
        
        print(f"Deposit transaction sent: {tx_hash}. Waiting for confirmation...")
        w3.eth.wait_for_transaction_receipt(tx_hash)

        # 4. Sign the Intent ID for Authorization
        print("Signing intent for authorization...")
        message = encode_defunct(text=f"AgentBureau Intent: {intent_id}")
        signed_message = w3.eth.account.sign_message(message, private_key=PRIVATE_KEY)
        signature = signed_message.signature.hex()
        
        # 5. Retry with payment signature and authorization
        print("Retrying with payment signature and authorization...")
        retry_headers = {
            "Idempotency-Key": idempotency_key,
            "PAYMENT-SIGNATURE": tx_hash,
            "PAYMENT-AUTHORIZATION": signature
        }
        final_response = httpx.post(SERVICE_ENDPOINT, json=payload, headers=retry_headers)
        
        if final_response.status_code == 200:
            print("Success!")
            print(final_response.json())
        else:
            print(f"Final call failed with status {final_response.status_code}")
            print(final_response.text)
    else:
        print(f"Status: {response.status_code}")
        print(response.text)

if __name__ == "__main__":
    run_flow()
